File: ruby/lib/game_machine/systems/request_handler.rb
module GameMachine
  module Systems
    class RequestHandler < Actor

      def on_receive(message)
        if message.is_a?(ClientMessage)
          if message.has_player_logout
            logged_out(message)
          elsif message.has_client_disconnect
            disconnected(message)
          elsif message.has_player
            update_entities(message)
            authenticate_player(message)
          else
            unhandled(message)
          end
        else
          unhandled(message)
        end
      end

      private

      def logged_out(message)
        PlayerRegistry.player_logout(
          message.player_logout.player_id
        )
      end

      def disconnected(message)
        PlayerRegistry.client_disconnect(
          message.client_disconnect.client_connection
        )
      end

      def update_entities(message)
        if message.get_entity_list
          message.get_entity_list.each do |entity|
            entity.set_player(message.player)
          end
        end
      end

      def authenticate_player(message)
        AuthenticationHandler.find_distributed_local(
          message.player.id
        ).tell(message,self)
      end

    end
  end
end

File: server/java/project/src/main/java/com/game_machine/routing/RequestHandler.java
package com.game_machine.routing;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import GameMachine.Messages.ChatDestroy;
import GameMachine.Messages.ClientConnection;
import GameMachine.Messages.ClientManagerRegister;
import GameMachine.Messages.ClientManagerUnregister;
import GameMachine.Messages.ClientMessage;
import GameMachine.Messages.Entity;
import akka.actor.ActorSelection;
import akka.actor.UntypedActor;

import com.game_machine.authentication.Authentication;
import com.game_machine.core.ActorUtil;

public class RequestHandler extends UntypedActor {

	private static final Logger logger = LoggerFactory.getLogger(RequestHandler.class);

	public static String name = "request_handler";
	private EntityRouter entityRouter;

	public RequestHandler() {
		entityRouter = new EntityRouter();
	}

	@Override
	public void onReceive(Object message) {
		if (message instanceof ClientMessage) {
			ClientMessage clientMessage = (ClientMessage) message;
			if (clientMessage.getEntityCount() >= 1) {
				List<Entity> entities = clientMessage.getEntityList();
				entityRouter.route(entities, clientMessage.getPlayer());
			}
		} else {
			unhandled(message);
		}
	}

	public static void unregisterClient(ClientMessage clientMessage) {
		logger.debug("unregister client " + clientMessage.getPlayer().getId());
		String playerId = clientMessage.getPlayer().getId();
		String clientId = clientMessage.getClientConnection().getId();
		Authentication.unregisterPlayer(playerId);
		ClientManagerUnregister unregister = new ClientManagerUnregister();
		unregister.setRegisterType("client");
		unregister.setName(clientId);
		Entity entity = new Entity();
		entity.setId(clientId);
		entity.setClientManagerUnregister(unregister);
		entity.setClientConnection(clientMessage.clientConnection);
		entity.setPlayer(clientMessage.getPlayer());
		ActorSelection sel = ActorUtil.getSelectionByName("GameMachine::ClientManager");
		sel.tell(entity, null);

		ChatDestroy chatDestroy = new ChatDestroy();
		chatDestroy.setPlayerId(playerId);
		sel = ActorUtil.getSelectionByName("GameMachine::GameSystems::ChatManager");
		sel.tell(chatDestroy, null);
	}

	public static void registerClient(ClientMessage clientMessage) {
		logger.debug("register client " + clientMessage.getPlayer().getId());
		ClientConnection clientConnection = clientMessage.getClientConnection();
		String clientId = clientConnection.getId();
		ClientManagerRegister register = new ClientManagerRegister();
		register.setRegisterType("client");
		register.setName(clientId);
		Entity entity = new Entity();
		entity.setId(clientId);
		entity.setClientManagerRegister(register);
		entity.setClientConnection(clientConnection);
		entity.setPlayer(clientMessage.getPlayer());
		ActorUtil.ask("GameMachine::ClientManager", entity, 50);
	}

}
